overlord/servicestate/servicestate.go

```diff
@@ -49,9 +49,42 @@ type Instruction struct {
 
 type ServiceActionConflictError struct{ error }
 
+type ResolvedAppInfo struct {
+	snap.AppInfo
+	ExplicitlyRequested bool
+}
+
+func NewResolvedAppInfo(app *snap.AppInfo) *ResolvedAppInfo {
+	resolvedAppInfo := new(ResolvedAppInfo)
+	resolvedAppInfo.AppInfo = *app
+	return resolvedAppInfo
+}
+
+type ResolvedAppInfosByAppName []*ResolvedAppInfo
+
+func NewResolvedAppInfosByAppName(appInfos []*snap.AppInfo) ResolvedAppInfosByAppName {
+	resolvedAppInfos := make(ResolvedAppInfosByAppName, 0, len(appInfos))
+	for _, app := range appInfos {
+		resolvedAppInfos = append(resolvedAppInfos, NewResolvedAppInfo(app))
+	}
+	return resolvedAppInfos
+}
+
+func (a ResolvedAppInfosByAppName) Len() int      { return len(a) }
+func (a ResolvedAppInfosByAppName) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ResolvedAppInfosByAppName) Less(i, j int) bool {
+	iName := a[i].Snap.InstanceName()
+	jName := a[j].Snap.InstanceName()
+	if iName == jName {
+		return a[i].Name < a[j].Name
+	}
+	return iName < jName
+}
+
 // serviceControlTs creates "service-control" task for every snap derived from appInfos.
-func serviceControlTs(st *state.State, appInfos []*snap.AppInfo, inst *Instruction) (*state.TaskSet, error) {
+func serviceControlTs(st *state.State, appInfos []*ResolvedAppInfo, inst *Instruction) (*state.TaskSet, error) {
```

Comment:
the slightly more cumbersome approach that doesn't need a new type is to add a new argument that is a set of the explicitly mentioned services  explicitApps map[string]bool

Reply:
I actually went for another approach, possibly even simpler: since the list of explicit services is computed out of the snap info and of the inst.Names array, I've added a method which does exactly this computation and I call it only from serviceControlTs().